### PR TITLE
feat(cli): support for transformers

### PR DESCRIPTION
**Pull request by wkeylin**
fix: #1369

### CODE REVIEW

**Comment by antfu** [comment]
Can we have some tests for it? Thanks

**Comment by wkeylin** [comment]
> Can we have some tests for it? Thanks

Please review again

File: test/cli.test.ts [hunk 1]
```diff
@@ -37,6 +37,32 @@ export default defineConfig({
...
+export default defineConfig({
+  transformers: [transformerVariantGroup()]
+})
+    `.trim(),
+    })
+    expect(output).toMatchSnapshot()
+  })
+
+  it('supports directives transformer', async () => {
+    const { output } = await runCli({
```
Comment (chris-zhu) [inline test/cli.test.ts:54]: `runCli` only output `uno.css` file content. I would prefer to see the tranformed code.

File: packages/cli/src/index.ts [hunk 2]
```diff
@@ -99,12 +99,38 @@ export async function build(_options: CliOptions) {
...
   async function generate(options: ResolvedCliOptions) {
+    const sourceCache = Array.from(fileCache).map(([id, code]) => ({ id, code }))
+
     const outFile = resolve(options.cwd || process.cwd(), options.outFile ?? 'uno.css')
-    const { css, matched } = await uno.generate(
-      [...fileCache].join('\n'),
+
+    const preTransform = await transformFiles(sourceCache, 'pre')
+    const defaultTransform = await transformFiles(preTransform)
+    const postTransform = await transformFiles(defaultTransform, 'post')
```
Comment (chris-zhu) [inline packages/cli/src/index.ts:121]: I personally think it's unnecessary, `enfore` is more used as a plugin property, like `vite`

**Comment by chris-zhu** [comment]
@wkeylin BTW, Can you take a look at this pr #1602 ? maybe merged together.

**Comment by wkeylin** [comment]
>

File: packages/cli/src/index.ts [hunk 3]
```diff
@@ -99,12 +99,38 @@ export async function build(_options: CliOptions) {
...
+    await Promise.all(
+      postTransform.filter(({ transformedCode }) => !!transformedCode)
+        .map(({ transformedCode, id }) => new Promise<void>((resolve) => {
+          if (existsSync(id))
+            fs.writeFile(id, transformedCode as string, 'utf-8').then(resolve)
+        })),
+    )
+
+    const { css, matched } = await ctx.uno.generate(
+      [...postTransform.map(({ code, transformedCode }) => transformedCode ?? code)].join('\n'),
```
Comment (wkeylin) [inline packages/cli/src/index.ts:133]: I have resolved this problem #1602 in my code @chris-zhu

File: packages/cli/src/index.ts (hunk 2 above)
Comment (wkeylin) [reply packages/cli/src/index.ts:121]: But the `transformerVariantGroup`, `transformerAttributifyJsx`, `transformerCompileClass` default enforce is `pre`, others transformers default enforce is `undefined`, `applyTransformers` will filter the corresponding transformer

Comment (chris-zhu) [reply packages/cli/src/index.ts:121]: You are right. 😅

**Review by chris-zhu — APPROVED** [review]

**Pull request merged**